src/cli/cli.cpp

```diff
@@ -1453,6 +1453,20 @@ template <> otError Interpreter::Process<Cmd("delaytimermin")>(Arg aArgs[])
 }
 #endif
 
+template <> otError Interpreter::Process<Cmd("detach")>(Arg aArgs[])
+{
+    otError error = OT_ERROR_NONE;
+
+    SuccessOrExit(error = otThreadDetachGracefully(GetInstancePtr(), &Interpreter::HandleDetachGracefullyResult, this));
+
+    mDetachAsync = (aArgs[0] == "async");
```

Comment:
Currently both `detach` and `detach async` are for tests only.

Reply:
sgtm to have detach async. 

Even though I prefer slightly to remove it if it's not used. 

`ping async` was added merely because OTNS is using async pings a lot.

Reply:
I'm inclined to remove `detach async` which is just for testing. On the other hand, we may return `busy` error trying `thread stop` when detaching. Thoughts?

Reply:
`detach` is also for test only, because the actual use is through the C API.

We have 2 choices:

1. Make `otThreadSetEnabled(false)` finalize the detaching. This is the current implementation.
2. Disallow `otThreadSetEnabled(false)` while detaching.

I think they are mostly the same regarding complexity. I'm inclined to 1 as it doesn't change the behaviors of existing API. Thoughts?

Reply:
I don't think detach is for test only. Users can actually use this command to detach gracefully. For example, in project otbr, the web page is calling cli commands instead of c APIs. I fine with either choices of `otThreadSetEnabled(false)`.

Reply:
Considering future use, I think `detach async` can be used when the user doesn't care about when the detaching actually finishes, like what we do when turning off WiFi on phone UI.

Also, if we keep the current implementation of `otThreadSetEnabled(false)`, we need `detach async` as we don't have any other way to test it.

Reply:
> Disallow otThreadSetEnabled(false) while detaching.

I think it adds additional restrictions that are not necessary. 

> I'm inclined to remove detach async which is just for testing.

+1. Inclined to remove it unless it's already used by some GitHub CI or test cases.

Reply:
> Also, if we keep the current implementation of otThreadSetEnabled(false), we need detach async as we don't have any other way to test it.

Agree. Fine to me to keep `detach async` if we want to use it for test.

Reply:
BTW, I am against adding `async` modes for commands because we may implement async mode for all commands in a consistent way similar to `bash` like below:
```
> detach & # run detach in the background
> thread stop
Done
< detach Done
```

Reply:
A consistent way is good. We can remove the async command after it's done.

Reply:
Fine to keep it. Should we document the `otThreadSetEnabled(false)` impact on detaching gracefully? Like it will break the detaching process.

Reply:
Added documentation to `otThreadSetEnabled`.